File: service/src/api/websocket/game-events.ts
import type { InboundGameEvent, OutboundGameEvent } from "@takes-a-village/shared";

import type { Game } from "../../game/game.js";

export interface GameEventMessenger {
  broadcast(gameId: string, packet: OutboundGameEvent): void;
  sendPersonal(gameId: string, userId: string, packet: OutboundGameEvent): void;
  broadcastStates(game: Game): void;
}

type RoutedGameEvent = Exclude<InboundGameEvent, { event: "join_room" }>;

export function processGameEvent(
  game: Game,
  userId: string,
  packet: RoutedGameEvent,
  messenger: GameEventMessenger,
): void {
  switch (packet.event) {
    case "start_game_request":
      if (game.hostId === userId && game.startGame()) {
        messenger.broadcast(game.id, { event: "game_started", data: { day: 1 } });
        messenger.broadcastStates(game);
      }
      return;
    case "request_update": {
      const state = game.getStateForPlayer(userId);
      if (state) messenger.sendPersonal(game.id, userId, { event: "game_state", data: state });
      return;
    }
    case "send_chat":
      deliverChatEvent(game, userId, packet.data, messenger);
      return;
    case "create_chat":
      if (game.createChat(userId, packet.data.name, packet.data.memberIds)) messenger.broadcastStates(game);
      return;
    case "submit_action": {
      if (game.status === "WAITING") return;
      const player = game.players.get(userId);
      if (packet.data.action_command === "FINISH_PHASE" && player?.finishedPhase) return;
      const accepted = game.handleAction(userId, {
        action_command: packet.data.action_command,
        payload: packet.data.payload,
      });
      if (accepted) messenger.broadcastStates(game);
      else messenger.sendPersonal(game.id, userId, {
        event: "error",
        data: {
          message: "Action rejected by game rules.",
          action_command: packet.data.action_command,
        },
      });
    }
  }
}

export function deliverChatEvent(
  game: Game,
  userId: string,
  data: { content: string; to_id: string },
  messenger: GameEventMessenger,
): boolean {
  const message = game.handleChat(userId, data.content, data.to_id);
  if (!message) return false;
  const packet = { event: "new_chat_message", data: message } satisfies OutboundGameEvent;
  if (message.to_id === "GLOBAL") {
    messenger.broadcast(game.id, packet);
    return true;
  }

  const group = game.chats.find((chat) => chat.id === message.to_id);
  const recipients = group?.member_ids ?? [message.from_id, message.to_id];
  for (const recipientId of new Set(recipients)) {
    if (recipientId) messenger.sendPersonal(game.id, recipientId, packet);
  }
  return true;
}

File: service/api/websocket/game_events.py
async def _send_queued_notifications(game, game_id, manager):
    if not hasattr(game, "drain_notifications"):
        return
    for player_id in list(getattr(game, "players", {}).keys()):
        for notification in game.drain_notifications(player_id):
            await manager.send_personal_message({
                "event": "game_notification",
                "data": notification,
            }, game_id, player_id)


async def process_game_event(event, payload, game_id, user_id, game, manager):
    if event == "start_game_request":
        if game.host_id == user_id and game.start_game():
            await manager.broadcast_to_game(
                {"event": "game_started", "data": {"day": 1}}, game_id)
            await manager.broadcast_game_state(game_id, game)
    elif event == "request_update":
        await manager.send_personal_message(
            {"event": "game_state", "data": game.get_state_for_player(user_id)},
            game_id, user_id)
    elif event == "send_chat":
        message = game.handle_chat(user_id, payload)
        if message:
            data = message.to_dict()
            group_chat = next(
                (chat for chat in game.chats if chat.id == message.to_id), None)
            if message.to_id == "GLOBAL":
                await manager.broadcast_to_game(
                    {"event": "new_chat_message", "data": data}, game_id)
            elif group_chat:
                packet = {"event": "new_chat_message", "data": data}
                for member_id in group_chat.member_ids:
                    await manager.send_personal_message(
                        packet, game_id, member_id)
            else:
                packet = {"event": "new_chat_message", "data": data}
                await manager.send_personal_message(packet, game_id, message.from_id)
                await manager.send_personal_message(packet, game_id, message.to_id)
    elif event == "submit_action":
        if game.status == "WAITING":
            return
        command = payload.get("action_command", payload.get("actionId"))
        player = game.players.get(user_id)
        if command == "FINISH_PHASE" and player and player.finished_phase:
            return
        if game.handle_action(user_id, payload):
            await _send_queued_notifications(game, game_id, manager)
            await manager.broadcast_game_state(game_id, game)
        else:
            await manager.send_personal_message({
                "event": "error",
                "data": {"message": "Action rejected by game rules.",
                         "action_command": command},
            }, game_id, user_id)
    elif event == "create_chat":
        if game.create_chat(user_id, payload["name"], payload["memberIds"]):
            await manager.broadcast_game_state(game_id, game)
